splitgpg2/__init__.py

```diff
@@ -0,0 +1,804 @@
+            except ValueError:
+                untrusted_inq = untrusted_args
+                untrusted_inq_args = None
+            try:
+                inquire = self.expected_inquires[untrusted_inq]
+            except KeyError:
+                raise Filtered
+            await inquire(untrusted_args=untrusted_inq_args)
+        except Exception as e:  # pylint: disable=broad-except
+            # propagate exception back to self.send_command
+            if not self.agent_responded.done():
+                self.agent_responded.set_exception(e)
+
+    # region INQUIRE commands sent from gpg-agent
+    #
+
+    async def inquire_KEYPARAM(self, untrusted_args):
+        if untrusted_args:
+            self.server.abort('unexpected arguments to KEYPARAM inquire')
+        await self.server.send_inquire(b'KEYPARAM', {
+            b'D': self.inquire_command_D,
+            b'END': self.inquire_command_END,
+        })
+
+    async def inquire_PINENTRY_LAUNCHED(self, untrusted_args):
+        args = untrusted_args
+
+        await self.server.send_inquire(b'PINENTRY_LAUNCHED ' + args, {
+            b'END': self.inquire_command_END,
+        })
+
+    async def inquire_CIPHERTEXT(self, untrusted_args):
+        if untrusted_args:
+            self.server.abort('unexpected arguments to CIPHERTEXT inquire')
+        await self.server.send_inquire(b'CIPHERTEXT', {
+            b'D': self.inquire_command_D,
+            b'END': self.inquire_command_END,
+        })
+
+    async def inquire_NEWPASSWD(self, untrusted_args):
+        if untrusted_args:
+            self.server.abort('unexpected arguments to NEWPASSWD inquire')
+        await self.server.send_inquire(b'NEWPASSWD', {
+            b'D': self.inquire_command_D,
+            b'END': self.inquire_command_END,
+        })
+
+    # endregion
+
+    # region INQUIRE responses sent by client back to the agent
+    #
+    # each function returns whether further responses are expected
+
+    async def inquire_command_D(self, *, untrusted_args):
+        # XXX: should we sanitize this here?
+        self.agent_write(b'D ' + untrusted_args + b'\n')
+        return True
+
+    async def inquire_command_END(self, *, untrusted_args):
+        if untrusted_args:
+            self.server.abort('unexpected arguments to END')
+        self.agent_write(b'END\n')
+        return False
+
+    # endregion
+
+
+TIMER_NAMES = {
+    'QUBES_SPLIT_GPG2_PKSIGN_AUTOACCEPT_TIME': 'PKSIGN',
+    'QUBES_SPLIT_GPG2_PKDECRYPT_AUTOACCEPT_TIME': 'PKDECRYPT',
+}
+
+
+def main():
+    # request bi-directional socket on stdin
+    if 'QREXEC_AGENT_PID' in os.environ:
+        try:
+            os.kill(int(os.environ['QREXEC_AGENT_PID']), signal.SIGUSR1)
+        except ProcessLookupError:
+            pass
+
+    client_domain = os.environ['QREXEC_REMOTE_DOMAIN']
+    loop = asyncio.get_event_loop()
+    sock = socket.fromfd(sys.stdin.fileno(), socket.AF_UNIX, socket.SOCK_STREAM)
```

Comment:
Is there an advantage compared to using stdin/stdout?

Reply:
This will fit nicely connected with https://github.com/QubesOS/qubes-core-qrexec/pull/3

Reply:
Yes, qrexec-agent set it as a socket.